### PR TITLE
Delegation in user profile

### PR DESCRIPTION
**Pull request by fisx**
related: #652, #651, #682

user profile has a list of buttons now, one for each class and one for school, all of them toggle.

### CODE REVIEW

**Comment by fisx** [comment]
i don't understand the error in the test suite.  @andorp can you easily see what's going on there?  i thought that path should not return 405?  perhaps something about an invalid user id in the path?  (seems unlikely to me, though.)

anyway.  giving up for today!

File: src/Frontend/Path.hs [hunk 1]
```diff
@@ -736,6 +736,8 @@ isPostOnly = \case
       case m of
           UserDelegateVoteOnSchoolSpace    -> True
           (UserDelegateVoteOnClassSpace _) -> True
+          UserWithdrawDelegationOnSchoolSpace    -> True
+          (UserWithdrawDelegationOnClassSpace _) -> True
```
Comment (fisx) [inline src/Frontend/Path.hs]: D'OH!  (-:

thanks...

File: src/Frontend/Path.hs [hunk 2]
```diff
@@ -25,7 +25,7 @@ module Frontend.Path
     ( HasPath(..)
     , Top(..)
     , AllowedMethod(..)
-    , Main
+    , Main(Broken)
```
Comment (andorp) [inline src/Frontend/Path.hs]: Please hide the `Broken` constructor.

File: src/Frontend/Page/User.hs [hunk 3]
```diff
@@ -291,14 +324,14 @@ delegatedVotesClass userId = do
 
 delegatedVotes :: (ActionPersist m, ActionUserHandler m)
       => AUID User -> DScope -> m PageUserProfileDelegatedVotes
-delegatedVotes userId scope = do
+delegatedVotes userId _scope = do  -- TODO: about dropping _scope on the floor here, see #682.  see also: userDelegateeListsMap
```
Comment (andorp) [inline src/Frontend/Page/User.hs]: Leftover `TODO` item.

File: src/Action.hs [hunk 4]
```diff
@@ -867,8 +858,12 @@ eventLogUserVotesOnComment ck@(CommentKey _ ideaId parentIds _) v = do
 
 eventLogUserDelegates ::
       (ActionUserHandler m, ActionPersist m, ActionCurrentTimestamp m, ActionLog m)
-      => DScope -> AUID User -> m ()
-eventLogUserDelegates scope delegateId = do
+      => Bool -> DScope -> AUID User -> m ()
+eventLogUserDelegates setOrWithdraw scope delegateId = do
```
Comment (andorp) [inline src/Action.hs]: I would introduce two functions on top of this one. The two would be `eventLogUserSetsDelegate` and `eventLogUserWithdrawsDelegate`.

File: src/Frontend/Path.hs (hunk 2 above)
Comment (fisx) [inline src/Frontend/Path.hs]: why?  won't do it!  (-:

i don't see any benefit (it's called "broken" and shouldn't even be locally defined in this module, and i was confused by the fact it was hidden and have wasted some time figuring out why.)

File: src/Frontend/Page/User.hs (hunk 3 above)
Comment (fisx) [inline src/Frontend/Page/User.hs]: there are a few more.  i would like to resolve them separately.

File: src/Action.hs (hunk 4 above)
Comment (fisx) [inline src/Action.hs]: good point.  will do.

**Pull request merged**